invoice/invoice.go

```
@@ -0,0 +1,966 @@
+	// address.
+	segwitPrefix := hrp[2:4] + "1"
+	if !chaincfg.IsBech32SegwitPrefix(segwitPrefix) {
+		return nil, fmt.Errorf("unknown network \"%v\"", hrp[2:4])
+	}
+
+	// Extract which network this invoice is meant for.
+	var net *chaincfg.Params
+	switch hrp[2:4] {
+	case "bc":
+		net = &chaincfg.MainNetParams
+	case "tb":
+		net = &chaincfg.TestNet3Params
+	case "sb":
+		net = &chaincfg.SimNetParams
+	default:
+		return nil, fmt.Errorf("unknown network: %v", hrp[2:4])
+	}
+	decodedInvoice.Net = net
+
+	// Optionally, if there's anything left of the HRP, it encodes the
+	// payment amount.
+	if len(hrp) > 4 {
+		amount, err := decodeAmount(hrp[4:])
+		if err != nil {
+			return nil, err
+		}
+		decodedInvoice.MilliSat = &amount
+	}
+
+	// Everything except the last 520 bits of the data encodes the invoice's
+	// timestamp and tagged fields.
+	invoiceData := data[:len(data)-104]
+
+	// Parse the timestamp and tagged fields, and fill the Invoice struct.
+	if err := parseData(&decodedInvoice, invoiceData, net); err != nil {
+		return nil, err
+	}
+
+	// The last 520 bits (104 groups) make up the signature.
+	sigBase32 := data[len(data)-104:]
+	sigBase256, err := bech32.ConvertBits(sigBase32, 5, 8, true)
+	if err != nil {
+		return nil, err
+	}
+	var sigBytes [64]byte
+	copy(sigBytes[:], sigBase256[:64])
+	recoveryID := sigBase256[64]
+
+	// The signature is over the hrp + the data the invoice, encoded in
+	// base 256.
+	taggedDataBytes, err := bech32.ConvertBits(invoiceData, 5, 8, true)
+	if err != nil {
+		return nil, err
+	}
+
+	toSign := append([]byte(hrp), taggedDataBytes...)
+
+	// We expect the signature to be over the single SHA-256 hash of that
+	// data.
+	hash := chainhash.HashB(toSign)
+
+	// If pubkey was provided as a tagged field, use that to verify the
+	// signature, if not do public key recovery.
+	if decodedInvoice.PubKey != nil {
+		var signature *btcec.Signature
+		lnwire.DeserializeSigFromWire(&signature, sigBytes)
```

Looks like we're ignoring an error here.